try to give a clue if you cannot discard

fpvandoorn opened the pull request:
Also, if the newest clued card is definitely not playable, then it is treated as a safe hint.

Now players will almost never bomb if they cannot discard. If they cannot discard ...
- ... they will first try to clue 5s if there is no pile with a 4 on top (such clue will be interpreted as a safe hint)
- Otherwise they will clue 1 if all 1s have been played
- Otherwise they will clue a rank which the next player doesn't have in their hand (this search can be extended to suits or other players)
- Otherwise they will play their newest card (which is - I think - slightly better than playing a random card)

---

Note: I noticed one thing in a game which confused me for a while: a play hint can be given by a player, but which is interpreted as a safe hint by the next player (if they clue rank including the oldest unclued card).

Commit:
try to give a clue if you cannot discard

Also, if the newest clued card is definitely not playable, then it is treated as a safe hint

cangiuli commented:
> a play hint can be given by a player, but which is interpreted as a safe hint by the next player

This is definitely a bug. `givePlayHint` should also require that the hint is not a save hint, preferably by using a modified version of `isSaveClue`. (I can take a crack at that some time.)

> Otherwise they will play their newest card (which is - I think - slightly better than playing a random card)

That seems reasonable.

cangiuli commented:
I moved some code around, and defined a type `SSet` (yeah, I know) for sets of suits, and `RSet` for sets of ranks. I think these will make some of your code nicer. Soon I will make the `clued` functions use  these instead of lists. (Of course, lists are still needed for ordered information.) But unfortunately these changes will break your new code...

fpvandoorn commented:
> But unfortunately these changes will break your new code...

That's alright. Feel free to break as much as you need :-)

cangiuli commented:
Alright, take a look. I deleted `cluedNotRank` and `cluedNotSuit`, and replaced them with `possibleRanks` and a new version of `possibleSuits`. Both return sets instead of lists, and they look cleaner. I'm not sure that `cluedRank` and `cluedSuit` are still necessary, but I didn't delete them just in case.

In your pull request, I think the only affected function is `isNotPlayable`, which should get simpler.

cangiuli commented:
I should point out that the difference between `cluedSuit` and `possibleSuits` is that the latter considers both positive and negative color hints, while the former only considers positive color hints.

fpvandoorn commented:
I updated my commit. 

In `set-list.sml` I didn't find an analogue of `List.all`. To check "for all `x` in `A` we have `P x`" is there a better way than `List.all P (ListSet.toList A)`?

Commit:
discard least vital card instead of playing newest card

fpvandoorn commented:
Also, is there a way to print out only certain turns in the log, depending on whether some function was called during that turn? That would be helpful for debugging. 
Imperatively, I would add something like `globalVar := true` to the function, and let the trace be `fn _ => if globalVar then (* print *); globalVar := false else ()`. 
Is there a nice way to do this is SML? I suppose we could make `play` output something of type `action * bool`, but that is a bit messy.

I'm now working on a better action when you have nothing else to do, but that doesn't happen very often :-)

fpvandoorn commented:
Added one more commit.

cangiuli commented:
> To check "for all `x` in `A` we have `P x`" is there a better way than `List.all P (ListSet.toList A)`?

This is probably the right way to do it (although it's `SSet.toList`). If it comes up enough times we can add such a function to `UtilHanabi`. Note that `toList` is "free", since an `SSet` is just a sorted list anyway.

> Also, is there a way to print out only certain turns in the log, depending on whether some function was called during that turn? That would be helpful for debugging.

I should factor out some of the printing code so that the players have access to the printing functions (although of course it wouldn't include their own cards). Then you could just print the relevant turns from inside the player logic. A temporary solution is to print "LOOK HERE" in the player whenever the condition comes up, and just search the full game log for it.

fpvandoorn commented:
> Note that `toList` is "free"

good point

> Then you could just print the relevant turns from inside the player logic. 

Oh, that would be nice. It doesn't matter what the current player has to judge whether the move is the correct one.

> A temporary solution is to print "LOOK HERE"

Yes, I've indeed been doing that for now.

Commit:
main: print some statistics

Commit:
add current scores to README

Commit:
start on memory

Commit:
add turnNumber to state and fstate

also try to update isPlayable

Commit:
write function which outputs the state of the previous turn

This required to change the type of datatype play. Closes #7

Commit:
Fix bugs in util and make memory work

closes #6

Commit:
keep track of all playable cards in everyone's hand

Commit:
add *~ to .gitignore

Commit:
better initialization by giving the number of players as argument to play

fpvandoorn commented:
I added all my commits to this PR.

Commit:
check cards which are certainly playable or certainly not playable

Commit:
new datastructure for memory

Commit:
add advanced player

fpvandoorn commented:
I played around a bit more. I made the following changes:
* if you have a playable card, but the next player needs a save, you save first
* you can give play clues to players other than the next
* if a card is known to be playable (either since a play clue was given, or enough information is known about that card), no play clue will be given to any copy of that card
* I implemented all this in a new player `advanced.sml`. I removed the memory from `simple.sml`. The advanced player gets 1-2 more points than the simple player on average.

Commit:
advanced: save before giving play clue to later players, and give smarter play clues

cangiuli commented:
Can't we reconstruct turn number pretty straightforwardly from the rest of the game state? And are we using it anywhere?

fpvandoorn commented:
We can reconstruct the turn number from the length of the log. It is currently only used when printing the game state. We can remove it if you want.

cangiuli commented:
Even the players can reconstruct it, from the size of the deck / turns left, and number of players. That would be more efficient than checking the length of the log.

cangiuli commented on src/advanced.sml:
```diff
@@ -0,0 +1,416 @@
+     | Maybe => default
+
+  (* check whether a card in your own hand is definitely not vital *)
+  fun isNotVital (s : state) (is : info list) : bool =
```
This seems like it might be expensive...it might be better to implement this in a way that isn't as bad when `possibleSuits` or `possibleRanks` are the full sets.

fpvandoorn replied:
Yes, it's not very efficient. Do you have suggestions on how to improve it? It's hardly ever called though, only by `discardAny`, which is if you have to discard a marked card from your hand.

cangiuli commented on src/advanced.sml:
```diff
@@ -0,0 +1,416 @@
+    | isSaveClueToMe s _ = false
+
+  (* Is a rank hint which hints the cards in positions l interpreted as a save hint? *)
+  fun isSaveClue (s : state) (pl : player) (l : int list) : bool =
```
I guess this function isn't needed?

fpvandoorn replied:
I removed it in `simple.sml`, where it wasn't used indeed. I do use it in `advanced.sml`.

cangiuli commented on src/advanced.sml:
```diff
@@ -0,0 +1,416 @@
+    fn s => (
+    m := withPlayable (!m) (updatePlayable s (#playable (!m)));
+    (* print (memoryToString (!m) ^"\n"); *)
+    (case playPlayableCard (hd (#playable (!m))) of
```
`Option.map` performs a case like this. Then I think you can compose `Play` with this function and `playPlayableCard` isn't necessary at all.

fpvandoorn replied:
That's indeed nicer. It's now
```
   Option.map (fn i => giveSaveHint s otherwise (fn () => Play i))
               (Util.findIndex (fn b => b) (hd (#playable (!m))))
```

cangiuli commented on src/hanabi.sml:
```diff
@@ -295,16 +311,10 @@ struct
-               (case who (i-pl-1) of
-                     Me => (who i, HintedRank (Me,r,[]))
-                   | Other j => (who i, HintedRank (Other j,r,cs)))
+             Discarded x => (who i, Discarded x)
```
OK, I agree not separating `Me` and `Other` is better!

cangiuli commented on src/hanabi.sml:
```diff
@@ -330,19 +340,21 @@ struct
     in
-      if (trace (act,s'); gameOver s')
+      if (trace (act,s); gameOver s')
       then s'
```
If we're printing before the turn instead of after, then when the game ends shouldn't we `trace (act,s')` so the ending state is printed once?

cangiuli replied:
Actually, I think we should change where the newline is in the trace so it's clearer which action is associated to which state.

fpvandoorn replied:
Did you try my version? It is now:
```
state n
action by the player seeing state n

state n+1
action by the player seeing state n+1

```
This has my preference.

I added printing the end state by adding a `printState s` to `newGame` (which was how you printed the initial state previously). Then the last action isn't printed twice.

cangiuli replied:
Ah, okay. This seems fine.

fpvandoorn commented:
> Even the players can reconstruct it, from the size of the deck / turns left, and number of players. That would be more efficient than checking the length of the log.

You mean by counting how much clues have been given? That works most of the time, but not always, since it's possible to play a 5 without getting a clue back.

Commit:
print end state

Commit:
remove turnNumber

cangiuli commented:
No, `#turns s` says how many cards are left in the deck, or how many turns remain in the game. Since you know how many cards were in the deck on turn 1, you can figure out what turn it is, right?
